### ISSUE TITLE
Return type as generic base class is not allowed

### ISSUE THREAD

**Issue by umurgdk** [crystal-lang/crystal#2283]
I have a method which returns an object typed ```ParserThen(Array(Something), Array(Something))``` and its parent class is ```Parser(Array(Something))``` but compiler complains about its type. I simplified the problem.

https://carc.in/#/r/tut

```
Error in line 30: instantiating 'test()'

in line 23: type must be A(Int32), not C(Char, Int32)
```

```crystal
class A(T)
    def val
        [] of T
    end
end

class B(T) < A(T)
end

class C(T,U) < A(U)
    def initialize(@first : A(T), @second : A(U))
    end

    def first
        @first.val
    end

    def val
        @second.val
    end
end

def test : A(Int32)
    a = A(Char).new
    b = B(Int32).new

    C.new a, b
end

test
```

**Comment by asterite** [comment]
Hi,

Please provide the full code that triggers the error you are getting. I'm getting:

```
Error in ./foo.cr:1: undefined constant Parser
```

There's nothing we can do with that to help you.

**Comment by umurgdk** [comment]
@asterite I just simplified problem on the description and include a carc.in link to it.

**Comment by bcardiff** [comment]
The generics class model have some pitfalls. Inheriting from a generic class I think is the problematic thing currently. Composition instead of inheritance might offer a workaround.
 
https://carc.in/#/r/tv9
using `C.new(a, b) as A(Int32)`
leads to `can't cast to A(Int32) because it was never instantiated (Exception)`

https://carc.in/#/r/tvc
forcing a `foo = A(Int32).new`
leads to a `Error: you've found a bug in the Crystal compiler. Bug: trying to upcast A(Int32) <- C(Char, Int32)`

**Comment by bcardiff** [comment]
Might be related to #846

**Comment by asterite** [comment]
Yes, inheriting from a generic class will probably won't work in many cases, so for now I'd advice not to use that "feature". We'll probably fix that once type annotations of instance variables are mandatory. Chances of fixing this before that point are very low.

**Comment by asterite** [comment]
Closed in favor of #2665

**Comment by asterite** [comment]
Closed in favor of #2665
